src/core/qgsdataitem.cpp

```diff
@@ -1381,6 +1406,89 @@ QVector<QgsDataItem *> QgsFavoritesItem::createChildren( const QString &favDir,
   return children;
 }
 
+//-----------------------------------------------------------------------
+
+QgsBookmarksItem::QgsBookmarksItem( QgsDataItem *parent, const QString &name, QgsBookmarkManager *applicationManager, QgsBookmarkManager *projectManager )
+  : QgsDataCollectionItem( parent, name, QStringLiteral( "bookmarks:" ) )
+{
+  mType = Custom;
+  mApplicationManager = applicationManager;
+  mProjectManager = projectManager;
+  mIconName = QStringLiteral( "/mActionShowBookmarks.svg" );
+  populate();
+}
+
+QVector<QgsDataItem *> QgsBookmarksItem::createChildren()
+{
+  QVector<QgsDataItem *> children;
+  if ( mApplicationManager )
+    children << new QgsBookmarkManagerItem( this, tr( "User Bookmarks" ), mApplicationManager );
+  if ( mProjectManager )
+    children << new QgsBookmarkManagerItem( this, tr( "Project Bookmarks" ), mProjectManager );
+  return children;
+}
+
+QgsBookmarkManagerItem::QgsBookmarkManagerItem( QgsDataItem *parent, const QString &name, QgsBookmarkManager *manager )
+  : QgsDataCollectionItem( parent, name, QStringLiteral( "bookmarks:%1" ).arg( name.toLower() ) )
+{
+  mType = BookmarkManager;
+  mManager = manager;
+  mIconName = QStringLiteral( "/mIconFolder.svg" );
+  populate();
+}
+
+QVector<QgsDataItem *> QgsBookmarkManagerItem::createChildren()
+{
+  connect( mManager, &QgsBookmarkManager::bookmarkAdded, [ = ]( const QString & ) { refresh(); } );
+  connect( mManager, &QgsBookmarkManager::bookmarkChanged, [ = ]( const QString & ) { refresh(); } );
+  connect( mManager, &QgsBookmarkManager::bookmarkRemoved, [ = ]( const QString & ) { refresh(); } );
+
+  QVector<QgsDataItem *> children;
+  for ( const QString &group : mManager->groups() )
+  {
+    if ( group.isEmpty() )
+    {
+      for ( const QgsBookmark bookmark : mManager->bookmarksByGroup( QString( "" ) ) )
+      {
+        children << new QgsBookmarkItem( this, bookmark.name(), bookmark );
+      }
+    }
+    else
+    {
+      children << new QgsBookmarkGroupItem( this, group, mManager );
+    }
+  }
+  return children;
+}
+
+QgsBookmarkGroupItem::QgsBookmarkGroupItem( QgsDataItem *parent, const QString &name, QgsBookmarkManager *manager )
+  : QgsDataCollectionItem( parent, name, QStringLiteral( "bookmarks:%1" ).arg( name.toLower() ) )
+{
+  mType = Custom;
+  mManager = manager;
+  mIconName = QStringLiteral( "/mIconFolder.svg" );
+  populate();
+}
+
+QVector<QgsDataItem *> QgsBookmarkGroupItem::createChildren()
+{
+  QVector<QgsDataItem *> children;
+  for ( const QgsBookmark bookmark : mManager->bookmarksByGroup( mName ) )
+  {
+    children << new QgsBookmarkItem( this, bookmark.name(), bookmark );
+  }
+  return children;
+}
+
+QgsBookmarkItem::QgsBookmarkItem( QgsDataItem *parent, const QString &name, QgsBookmark bookmark )
+  : QgsDataItem( Bookmark, parent, name, QStringLiteral( "bookmarks:%1/%2" ).arg( bookmark.group().toLower(), bookmark.id() ) )
+{
+  mType = Bookmark;
+  mCapabilities |= Fertile;
```

Comment:
Hang on, we don't want the fertile flag here - that indicates that the item can have children

Reply:
Am I not removing the Fertile flag here?

Reply:
Nope, you're adding it ;)